### ISSUE TITLE
net/haproxy 'curves' : library does not support curve suite

### ISSUE THREAD

**Issue by waffshappen** [opnsense/plugins#1573]
**Describe the bug**
A clear and concise description of what the bug is, including last known working version (if any).
Last working: 19.7.5

In haproxy, with both openssl and libressl in previous versions the following in "advanced options: bind options pass-through" used to work:
`curves X25519:secp521r1:secp384r1`
(Same for 
`ecdhe X25519`
)
The same goes for any other combination, like "P-256" or just secp384r1 - even the manual string ( https://cbonte.github.io/haproxy-dconv/1.8/configuration.html#5.1-curves ) gives out the same error

But now, since the latest update, it instead throws:

`[ALERT] 311/102914 (36351) : parsing [/usr/local/etc/haproxy.conf:64] : 'bind *ip*:443' : 'curves' : library does not support curve suite`

**To Reproduce**
1. Enable the haproxy plugin & set " Enable HAProxy" to on in the settings
2. Create a virtual https public service
3. Enable advanced mode when editing
4. Input `curves X25519:secp521r1:secp384r1` into either Bind option pass-through or SSL option pass-through 
5. Save the config. Click test. Observe the error.
6. If saving and applying anyway observe haproxy start failure
7. Observe the much weaker P-256 curve being default if not applying this option

**Expected behavior**
Enabling the X25519 curve and setting it as primary server choice

**Comment by fraenki** [comment]
TBH, I have no idea where to look at, it's certainly not configurable in the HAProxy source package.
@fichtner Did you by any chance change OpenSSL options recently?

**Comment by fichtner** [comment]
We should clear if this is about LibreSSL or OpenSSL, but I suspect HAProxy 2 change is the culprit here.

**Comment by waffshappen** [comment]
> We should clear if this is about LibreSSL or OpenSSL, but I suspect HAProxy 2 change is the culprit here.

Very likely, as this happens with both, since .6, across two different installs. 
Atleast i noticed it in time as the automated letsencrypt haproxy restart let it fail overnight and it wouldnt work in the morning - which led me to discover this in the first place. ^^

But this should work in haproxy, even 2.x, according to the manual. (Unless they forgot to update the docs: https://cbonte.github.io/haproxy-dconv/2.0/configuration.html#5.1-curves )

**Comment by fraenki** [comment]
> Very likely, as this happens with both, since .6, across two different installs.

...but HAProxy 2.x was already included in OPNsense 19.7.5...

**Comment by fraenki** [comment]
So I've tested this on the OpenSSL version of OPNsense 19.7.6. When using ` curves X25519:secp521r1:secp384r1` I get the following error:

```
unable to set SSL curves list to 'X25519:secp521r1:secp384r1' 
```

However, when checking for supported options with `openssl ecparam -list_curves` and using `curves secp384r1:sect409r1` instead, everything is working as expected.

**Comment by waffshappen** [comment]
> So I've tested this on the OpenSSL version of OPNsense 19.7.6. When using ` curves X25519:secp521r1:secp384r1` I get the following error:
> 
> ```
> unable to set SSL curves list to 'X25519:secp521r1:secp384r1' 
> ```
> 
> However, when checking for supported options with `openssl ecparam -list_curves` and using `curves secp384r1:sect409r1` instead, everything is working as expected.

If you consult the openssl documentation
( https://wiki.openssl.org/index.php/Command_Line_Elliptic_Curve_Operations )
You'll find that the *25519 curves are a special case from the beginning, and are not listed in the ec/ecparam set. Nonetheless they are supported, work(ed) with the aforementioned X25519 curve specification in haproxy, and until the last update(s) did so on both open- and libressl.

However, on my libressl setup not even the `curves secp384r1` works. Both machines are running 19.7.6.

However, if .7.5 updated to HaProxy 2.x and was not a reboot-enforced update it could possibly be that i got the "last working" version number wrong and it was only in .7.4 - but it must be <30 days before my bug report, because thats the maximum time between haproxy restarts through letsencrypt.

**Comment by waffshappen** [comment]
Just confirmed, curves  with sect or secp is working on the openssl flavour, but not the X25519 case (as said, special case in both the openssl ec/ecparam stuff, and with haproxy - but in their official docs). 

The LibreSSL flavour doesnt support any manual curvesetting from haproxy, so possibly another, or atleast related bug.

**Comment by fraenki** [comment]
> However, if .7.5 updated to HaProxy 2.x and was not a reboot-enforced update it could possibly be that i got the "last working" version number wrong and it was only in .7.4

If you think this is a regression that was introduced in HAProxy 2.0, it might be worth to report a bug upstream:
https://github.com/haproxy/haproxy/issues

**Comment by waffshappen** [comment]
> > However, if .7.5 updated to HaProxy 2.x and was not a reboot-enforced update it could possibly be that i got the "last working" version number wrong and it was only in .7.4
> 
> If you think this is a regression that was introduced in HAProxy 2.0, it might be worth to report a bug upstream:
> https://github.com/haproxy/haproxy/issues

Regarding this, does
https://github.com/haproxy/haproxy/issues/366#issuecomment-555871558
give any insight/changes needed here?

**Comment by waffshappen** [comment]
> > However, if .7.5 updated to HaProxy 2.x and was not a reboot-enforced update it could possibly be that i got the "last working" version number wrong and it was only in .7.4
> 
> If you think this is a regression that was introduced in HAProxy 2.0, it might be worth to report a bug upstream:
> https://github.com/haproxy/haproxy/issues

A likely source has been found:
https://github.com/haproxy/haproxy/issues/366#issuecomment-557312000
And a patch (see the comment) has been released, can you try if it works please?

**Comment by fraenki** [comment]
@waffshappen Thanks for submitting and discussing this issue upstream! As Willy stated HAProxy 2.0.10 will be issued tomorrow and should include the fix. I'm pretty sure OPNsense 19.7.8 will include this fix.

**Comment by waffshappen** [comment]
Just updated and can confirm the issue has been fixed in 19.7.8!

(But didnt see it in the changelog)
